File: core/version_utils.py
"""Semantic version comparison for partner app force-update."""

from __future__ import annotations


def normalize_version(version: str) -> str:
    """Strip build metadata (e.g. 2.0.0+1 -> 2.0.0)."""
    if not version:
        return '0.0.0'
    v = str(version).strip()
    if '+' in v:
        v = v.split('+', 1)[0]
    if '-' in v and not v[0].isdigit():
        pass
    return v or '0.0.0'


def parse_version_parts(version: str) -> tuple[int, ...]:
    normalized = normalize_version(version)
    parts: list[int] = []
    for segment in normalized.split('.'):
        segment = segment.strip()
        if not segment:
            parts.append(0)
            continue
        num = ''
        for ch in segment:
            if ch.isdigit():
                num += ch
            else:
                break
        parts.append(int(num) if num else 0)
    while len(parts) < 3:
        parts.append(0)
    return tuple(parts[:3])


def compare_versions(left: str, right: str) -> int:
    """
    Compare two version strings.
    Returns -1 if left < right, 0 if equal, 1 if left > right.
    """
    a = parse_version_parts(left)
    b = parse_version_parts(right)
    if a < b:
        return -1
    if a > b:
        return 1
    return 0


def is_version_below(current: str, minimum: str) -> bool:
    return compare_versions(current, minimum) < 0

File: pest_99_partner_app/lib/core/utils/version_utils.dart
/// Semantic version comparison (e.g. 2.0.0 vs 0.1.0+1).
int compareVersions(String left, String right) {
  final a = _parseParts(left);
  final b = _parseParts(right);
  for (var i = 0; i < 3; i++) {
    if (a[i] < b[i]) return -1;
    if (a[i] > b[i]) return 1;
  }
  return 0;
}

bool isVersionBelow(String current, String minimum) {
  return compareVersions(current, minimum) < 0;
}

String normalizeVersion(String version) {
  var v = version.trim();
  if (v.contains('+')) {
    v = v.split('+').first;
  }
  return v.isEmpty ? '0.0.0' : v;
}

List<int> _parseParts(String version) {
  final normalized = normalizeVersion(version);
  final parts = normalized.split('.').map((segment) {
    final digits = StringBuffer();
    for (final ch in segment.trim().split('')) {
      if (ch.codeUnitAt(0) >= 48 && ch.codeUnitAt(0) <= 57) {
        digits.write(ch);
      } else {
        break;
      }
    }
    return int.tryParse(digits.toString()) ?? 0;
  }).toList();
  while (parts.length < 3) {
    parts.add(0);
  }
  return parts.take(3).toList();
}
